WordPress/openverse-catalog
# Refactor NYPL script to use ProviderDataIngester class

## Fixes


Fixes #590 by @stacimc, fixes #710 by @AetherUnbound 

## Description


This PR replaces #546. Changes in the NYPL script:
- use of the DataIngester class. NYPL requests a single image API endpoint to get each image's data. To enable delayed requests for _detail_ endpoints, I added `get_detail_json` method to the base class, but I think its API can be greatly improved.
- filetype detection.
- category is set as `photograph`  for items with `photograph` as their `genre`.
- creator is added when there is a single creator, not only when there's an array of creators.
- `date_created` added to `meta_data`.
- `topics` extracted and added to `meta_data`. I was wondering if we should combine `physicalDescription` and `topics` into a single string and save it as `description` to improve the search relevancy for NYPL.

## Testing Instructions

Run NYPL DAG - it should run without errors. You can stop the `pull_data` step early to review the data that is saved in the database.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
